bread/tests/test_public_visible.py

```diff
@@ -0,0 +1,99 @@
+import secrets
+import traceback
+
+import django.urls
+from django.contrib.admindocs.views import simplify_regex
+from django.template import TemplateDoesNotExist
+from django.test import Client, TestCase
+from django_celery_results.models import TaskResult
+from django_extensions.management.commands import show_urls
+
+from bread.contrib.reports.models import Report
+
+ROOT_URLPATTERNS = django.urls.get_resolver(None).url_patterns
+
+
+class TestAnonymousVisible(TestCase):
+    # get a list of urlpatterns
+    ALL_URL_NAMES = {
+        simplify_regex(x[1])
+        for x in show_urls.Command().extract_views_from_urlpatterns(ROOT_URLPATTERNS)
+    }
+    MODEL_OPERATIONS_REQUIRING_PK = {
+        "copy",
+        "delete",
+        "edit",
+        "read",
+    }
+
+    # (model_objects, url, additional_operations)
+    BREAD_MODELS_URLPATTERNS = (
+        (Report.objects.all(), "/reports/reports/report/", {"excel"}),
+        (TaskResult.objects.all(), "/bread/django_celery_results/taskresult/", set()),
+    )
+    EXCEPTED_URLPATTERNS = {
+        "/admin/login/",
+        "/bread/accounts/login/",
+        "/bread/accounts/password_reset/",
+        "/bread/accounts/password_reset/done/",
+        "/bread/accounts/reset/<uidb64>/<token>/",
+        "/bread/accounts/reset/done/",
+        "/bread/auth/password_reset/",
+        "/bread/auth/reset/<uidb64>/<token>/",  # password reset
+        "/bread/auth/password_reset/done/",
+        "/bread/auth/reset/done/",
+        # TODO: figure out the way to manage this view.
+        "/bread/preferences/global/<slug:section>",
+    }
+
+    def test_visible(self):
+        # add the uncompiled url to exception
+        excepted_urlpatterns = self.EXCEPTED_URLPATTERNS.copy()
+        url_intpk = {x for x in self.ALL_URL_NAMES if "<int:pk>" in x}
+
+        # update the exceptions for some pages that can be visible to the public
+        # and those pages with dynamic url regex
+        # and those out of the bread's scope (e.g., basxconnect)
+        url_names = self.ALL_URL_NAMES - excepted_urlpatterns - url_intpk
+
+        # manually add a list of urls with a dynamic subdirectory.
+        # (e.g., <int:pk>)
+        for url in url_intpk:
+            url_names.add(url.replace("<int:pk>", str(secrets.randbelow(100))))
```

Comment:
Can you generalize this a bit? I mean so that it works for any type of parameter. If you look at https://github.com/basxsoftwareassociation/bread/blob/main/bread/utils/urls.py#L274 you can see the possible type of parameters. It is also possible to have a parameter without a type, in that case the type is str.
We should also be able to handle any parameter name, not only "pk".
This will also solve the problem with the ```"/bread/preferences/global/<slug:section>"``` URL.

Reply:
I replaced all types except `int` with randomized strings. Would you suggest what I should be specifically putting into `UUID`, `slug`, `path`, and even `str`?

Reply:
For UUID you should generate a uuid with pythons ```uuid``` module. Otherwise I think djangos URL-parse will raise an error. For slug, path and str a random str should be okay as long as they are alphanumeric.